crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -0,0 +1,306 @@
+use std::cmp::Ordering;
+
+use ruff_db::files::{File, FilePath};
+
+use crate::db::Db;
+use crate::types::Type;
+
+use super::{
+    class_base::ClassBase, Class, ClassLiteralType, DynamicType, InstanceType, KnownClass,
+    KnownInstanceType, TodoType,
+};
+
+/// Return an [`Ordering`] that describes the canonical order in which two types should appear
+/// in an [`crate::types::IntersectionType`] or a [`crate::types::UnionType`].
+///
+/// Two unions with equal sets of elements always have elements ordered the same way in our
+/// representation. This helps reduce memory usage, and also makes it easier to answer the
+/// question of whether two unions are equal.
+///
+/// ## Why not just implement `Ord` on `Type`?
+///
+/// It would be fairly easy to slap `#[derive(PartialOrd, Ord)]` on `Type`. However, this would
+/// order types according to their Salsa ID. While this would mean that types would always be
+/// consistently ordered in any single run of red-knot, the order in which they would appear
+/// might vary between different runs of red-knot. Unless we implemented an entirely different
+/// order for display purposes, this would make it difficult to write mdtests, and would also
+/// be quite confusing for users.
+///
+/// Moreover, it doesn't really "make sense" for `Type` to implement `Ord` in terms of the
+/// semantics. There are many different ways in which you could plausibly sort a list of types;
+/// this is only one (somewhat arbitrary, at times) possible ordering.
+pub(super) fn sort_union_elements<'db>(
+    db: &'db dyn Db,
+    left: &Type<'db>,
+    right: &Type<'db>,
+) -> Ordering {
+    if left == right {
+        return Ordering::Equal;
+    }
+
+    match (left, right) {
+        (Type::Never, Type::Never) => Ordering::Equal,
```

Comment:
This branch and other branches where `left == right` seem unnecessary?

Reply:
Nice catch!